File: Leetcode 3001-3100/3006/3006.py
# Leetcode 3006: Find Beautiful Indices in the Given Array I
# https://leetcode.com/problems/find-beautiful-indices-in-the-given-array-i/
# Solved on 4th of January, 2026
class Solution:
    def beautifulIndices(self, s: str, a: str, b: str, k: int) -> list[int]:
        """
        Finds all "beautiful" indices in a string `s`.
        An index `i` is beautiful if `s[i:i+len(a)] == a` and there exists an index `j`
        such that `s[j:j+len(b)] == b` and `abs(i - j) <= k`.

        :param s: The main string to search within.
        :param a: The first substring to find.
        :param b: The second substring to find.
        :param k: The maximum allowed absolute difference between indices `i` and `j`.
        :return: A list of all beautiful indices `i`.
        """
        sLen = len(s)
        aLen = len(a)
        bLen = len(b)

        indicesA = []
        for i in range(sLen - aLen + 1):
            if s[i:i + aLen] == a:
                indicesA.append(i)

        indicesB = []
        for j in range(sLen - bLen + 1):
            if s[j:j + bLen] == b:
                indicesB.append(j)

        result = []
        bIndex = 0
        bCount = len(indicesB)

        for i in indicesA:
            while bIndex < bCount and indicesB[bIndex] < i - k:
                bIndex += 1

            if bIndex < bCount and abs(indicesB[bIndex] - i) <= k:
                result.append(i)

        return result

File: Leetcode 3001-3100/3006/3006.java
// Leetcode 3006: Find Beautiful Indices in the Given Array I
// https://leetcode.com/problems/find-beautiful-indices-in-the-given-array-i/
// Solved on 4th of January, 2026
import java.util.ArrayList;
import java.util.List;

class Solution {
    /**
     * Finds all "beautiful" indices in a given string `s`.
     * An index `i` is beautiful if `s` contains string `a` starting at `i`,
     * and there exists an index `j` such that `s` contains string `b` starting at `j`,
     * and `|i - j| <= k`.
     * @param s The main string to search within.
     * @param a The first substring to find.
     * @param b The second substring to find.
     * @param k The maximum allowed absolute difference between the starting indices of `a` and `b`.
     * @return A list of all beautiful indices.
     */
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> indicesA = new ArrayList<>();
        List<Integer> indicesB = new ArrayList<>();
        List<Integer> result = new ArrayList<>();
        
        int lenS = s.length();
        int lenA = a.length();
        int lenB = b.length();

        for (int i = 0; i <= lenS - lenA; i++) {
            if (s.startsWith(a, i)) {
                indicesA.add(i);
            }
        }

        for (int i = 0; i <= lenS - lenB; i++) {
            if (s.startsWith(b, i)) {
                indicesB.add(i);
            }
        }

        int j = 0;
        for (int i : indicesA) {
            while (j < indicesB.size() && indicesB.get(j) < i - k) {
                j++;
            }
            if (j < indicesB.size() && Math.abs(indicesB.get(j) - i) <= k) {
                result.add(i);
            }
        }

        return result;
    }
}